crates/interpreter/src/exec.rs

```diff
@@ -1423,6 +1426,188 @@ impl<'m> Thread<'m> {
+        let mem = match self.mem_slice(mem, m, i, n / 8, true) {
+            None => return Err(trap()),
+            Some(x) => x,
+        };
+        macro_rules! convert {
+            ($T:ident, $t:ident, $s:ident) => {{
+                let ptr = mem.as_mut_ptr() as *mut $s;
+                let ptr = unsafe {
+                    paste::paste! { [< Atomic $s:upper >]::from_ptr(ptr) }
+                };
+                Val::$T(ptr.load(Ordering::SeqCst) as $t)
+            }};
+        }
+
+        let c = match (t, n) {
+            (NumType::I32, 32) => convert!(I32, u32, u32),
+            (NumType::I64, 64) => convert!(I64, u64, u64),
+            (NumType::I32, 8) => convert!(I32, u32, u8),
+            (NumType::I32, 16) => convert!(I32, u32, u16),
+            (NumType::I64, 8) => convert!(I64, u64, u8),
+            (NumType::I64, 16) => convert!(I64, u64, u16),
+            (NumType::I64, 32) => convert!(I64, u64, u32),
+            _ => unreachable!(),
+        };
+        self.push_value(c);
+        Ok(())
+    }
+
+    #[cfg(feature = "threads")]
+    fn atomic_store(
+        &mut self, mem: &mut Memory<'m>, t: NumType, n: usize, m: MemArg,
+    ) -> Result<(), Error> {
+        let c = self.pop_value();
+        let i = self.pop_value().unwrap_i32();
+        let mem = match self.mem_slice(mem, m, i, n / 8, true) {
+            None => return Err(trap()),
+            Some(x) => x,
+        };
+
+        macro_rules! convert {
+            ($s:ident, $t:ident) => {{
+                let atomic_val;
+                unsafe {
+                    paste::paste! {
+                        atomic_val = [< Atomic $t:upper >]::from_ptr(mem.as_mut_ptr() as *mut $t);
+                    }
+                }
+                paste::paste! {
+                    atomic_val.store((c.[<unwrap_ $s>]() as $t).into(),Ordering::SeqCst)
+                }
+            }};
+        }
+
+        match (t, n) {
+            (NumType::I32, 32) => convert!(i32, i32),
+            (NumType::I64, 64) => convert!(i64, i64),
+            (NumType::I32, 8) => convert!(i32, i8),
+            (NumType::I32, 16) => convert!(i32, i16),
+            (NumType::I64, 8) => convert!(i64, i8),
+            (NumType::I64, 16) => convert!(i64, i16),
+            (NumType::I64, 32) => convert!(i64, i32),
+            _ => unreachable!(),
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "threads")]
+    fn atomic_unop(
+        &mut self, op: AtomicUnOp, mem: &mut Memory<'m>, t: NumType, n: usize, m: MemArg,
+    ) -> Result<(), Error> {
+        let c = self.pop_value();
+        let i = self.pop_value().unwrap_i32();
+        let mem = match self.mem_slice(mem, m, i, n / 8, true) {
+            None => return Err(trap()),
+            Some(x) => x,
+        };
+        let c = match (t, n) {
+            (NumType::I32, 32) => {
+                Val::I32(op.i32(mem.as_mut_ptr() as *mut i32, c.unwrap_i32() as i32) as u32)
```

Comment:
Maybe after the sign issue is fixed, it would be possible to use a macro like:

```rust
($n:ident, $u:ident) => (op.$n(mem.as_mut_ptr() as *mut $u, c.unwrap_$n() as $u)
```
(probably using paste for the unwrap)

Reply:
Yeah, looks great.